ibis/expr/types/relations.py

```diff
@@ -104,8 +99,76 @@ def f(  # noqa: D417
     return f
 
 
+# TODO(kszucs): cover it with tests
+# TODO(kszucs): should use (table, *args, **kwargs) instead to avoid interpreting
+# nested inputs
+def bind(table: TableExpr, value: Any, prefer_column=True) -> Iterator[ir.Value]:
+    """Bind a value to a table expression."""
+    if prefer_column and isinstance(value, (str, int)):
+        yield table.column(value)
+    elif isinstance(value, ValueExpr):
+        yield value
+    elif isinstance(value, TableExpr):
+        for name in value.columns:
+            yield value.column(name)
+    elif isinstance(value, Deferred):
+        yield value.resolve(table)
+    elif isinstance(value, Selector):
+        yield from value.expand(table)
+    elif isinstance(value, (tuple, list)):
+        for v in value:
+            yield from bind(table, v, prefer_column=prefer_column)
+    elif isinstance(value, dict):
+        for k, v in value.items():
+            for val in bind(table, v, prefer_column=prefer_column):
+                yield val.name(k)
+    elif isinstance(value, ops.Value):
+        # TODO(kszucs): from certain builders, like ir.GroupedTable we pass
+        # operation nodes instead of expressions to table methods, it would
+        # be better to convert them to expressions before passing them to
+        # this function
+        yield value.to_expr()
+    elif callable(value):
+        yield value(table)
+    else:
+        yield literal(value)
+
+
+def unwrap_aliases(values: Iterator[ir.Value]) -> Mapping[str, ir.Value]:
+    """
+    Unwrap aliases into a mapping of {name: expression}.
+    """
+    result = {}
+    for value in values:
+        node = value.op()
+        if node.name in result:
+            raise com.IntegrityError(
+                f"Duplicate column name {node.name!r} in result set"
+            )
+        if isinstance(node, ops.Alias):
+            result[node.name] = node.arg
+        else:
+            result[node.name] = node
+    return result
+
+
+def dereference_values(parents, values):
+    """Trace and replace fields from earlier relations in the hierarchy."""
```

Comment:
This needs some elaboration, because it underlies a ton of the changes to generated SQL.

For example:

1. Why is the replace valid?
2. In what cases is it **not** valid?